adapters/algorix/algorix.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	EndpointTemplate template.Template
+}
+
+// Builder builds a new instance of the Foo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	endpoint, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+	bidder := &adapter{
+		EndpointTemplate: *endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var adapterRequests []*adapters.RequestData
+	var errs []error
+
+	adapterRequest, err := a.makeRequest(request)
+	if err == nil {
+		adapterRequests = append(adapterRequests, adapterRequest)
+	} else {
+		errs = append(errs, err)
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	algorixExt := parseAlgoriXExt(request)
+
+	if algorixExt == nil {
+		return nil, &errortypes.BadInput{Message: "Invalid ExtImpAlgoriX value"}
+	}
+
+	endPoint, err := a.getEndPoint(algorixExt)
+	if err != nil {
+		return nil, err
+	}
+
+	preProcess(request)
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     endPoint,
+		Body:    reqBody,
+		Headers: headers,
+	}, nil
+}
+
+//parseAlgoriXExt parse AlgoriX Ext
+func parseAlgoriXExt(request *openrtb2.BidRequest) *openrtb_ext.ExtImpAlgorix {
+	var extImpAlgoriX openrtb_ext.ExtImpAlgorix
+	for _, imp := range request.Imp {
+		var extBidder adapters.ExtImpBidder
+		err := json.Unmarshal(imp.Ext, &extBidder)
+		if err != nil {
+			continue
+		}
+		err = json.Unmarshal(extBidder.Bidder, &extImpAlgoriX)
+		if err != nil {
+			continue
+		}
+		return &extImpAlgoriX
+	}
+	return nil
+}
+
+// getEndPoint get Endpoint Target
+func (a *adapter) getEndPoint(ext *openrtb_ext.ExtImpAlgorix) (string, error) {
+	endPointParams := macros.EndpointTemplateParams{SourceId: ext.Sid, AccountID: ext.Token}
+	return macros.ResolveMacros(a.EndpointTemplate, endPointParams)
+}
+
+// preProcess process Request check and modify banner imp
+func preProcess(request *openrtb2.BidRequest) {
+	for i := range request.Imp {
+		if request.Imp[i].Banner != nil {
+			banner := *request.Imp[i].Banner
+			if (banner.W == nil || banner.H == nil || *banner.W == 0 || *banner.H == 0) && len(banner.Format) > 0 {
+				firstFormat := banner.Format[0]
+				bannerCopy := *request.Imp[i].Banner
```

Comment:
You already made a copy with the previous line `banner := *request.Imp[i].Banner`. Two copies are not necessary. Please just use one.